component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -2704,4 +2807,228 @@ private boolean isOTPMismatched(AuthenticationContext context) {
+        }
+        return EmailOTPAuthenticatorConstants.EMAIL_OTP_UPPER_CASE_ALPHABET_CHAR_SET
+                + EmailOTPAuthenticatorConstants.EMAIL_OTP_NUMERIC_CHAR_SET;
+    }
+
+    /**
+     * Append the recaptcha related params if recaptcha is enabled for Email OTP.
+     *
+     * @param request       HttpServletRequest
+     * @return string with the appended recaptcha params
+     */
+    private String getCaptchaParams(HttpServletRequest request, AuthenticationContext context) {
+
+        String captchaParams = "";
+
+        EmailOTPCaptchaConnector emailOTPCaptchaConnector = new EmailOTPCaptchaConnector();
+        emailOTPCaptchaConnector.init(EmailOTPServiceDataHolder.getInstance().getIdentityGovernanceService());
+        try {
+            if (emailOTPCaptchaConnector.isEmailRecaptchaEnabled(request) && isEmailOTPAsFirstFactor(context)) {
+                captchaParams = "&reCaptcha=true";
+            }
+        } catch (CaptchaException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Failed to determine if recaptcha for Email OTP is enabled", e);
+            }
+        }
+
+        return captchaParams;
+    }
+
+    /**
+     * This method checks if all the authentication steps up to now have been performed by authenticators that
+     * implements AuthenticationFlowHandler interface. If so, it returns true.
+     * AuthenticationFlowHandlers may not perform actual authentication though the authenticated user is set in the
+     * context. Hence, this method can be used to determine if the user has been authenticated by a previous step.
+     *
+     * @param context   AuthenticationContext
+     * @return true if all the authentication steps up to now have been performed by AuthenticationFlowHandlers.
+     */
+    private boolean isPreviousIdPAuthenticationFlowHandler(AuthenticationContext context) {
+
+        boolean hasPreviousAuthenticators = false;
+        Map<String, AuthenticatedIdPData> currentAuthenticatedIdPs = context.getCurrentAuthenticatedIdPs();
+        if (currentAuthenticatedIdPs != null && !currentAuthenticatedIdPs.isEmpty()) {
+            for (Map.Entry<String, AuthenticatedIdPData> entry : currentAuthenticatedIdPs.entrySet()) {
+                AuthenticatedIdPData authenticatedIdPData = entry.getValue();
+                if (authenticatedIdPData != null) {
+                    List<AuthenticatorConfig> authenticators = authenticatedIdPData.getAuthenticators();
+                    if (authenticators != null) {
+                        for (AuthenticatorConfig authenticator : authenticators) {
+                            hasPreviousAuthenticators = true;
+                            if (!(authenticator.getApplicationAuthenticator() instanceof AuthenticationFlowHandler)) {
+                                return false;
+                            }
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
We can use the streams API and use the `filter(Objects::nonNull)` to avoid these null checks

Reply:
It would make the code more readable as well :)